editoast: send correct 'start_time' for paced train conflicts

woshilapin opened the pull request:
Fixes #11705 

When sending paced train occurrences to `core`, asking for conflicts, the `start_time` was always the one from the base occurrence. This PR calculates the correct `start_time` for each occurrence.

The fix is relatively simple (last commit), but I made 3 preliminary commits about renaming some variables. Most of the code is for adding tests:
- a test in `editoast` to check the construction of the `core` request for conflicts
- an integration test, checking that a paced train:
  - always have conflicts with its own occurrences if the interval is very small
  - never have conflicts with its own occurrences if the interval is big enough

woshilapin committed:
editoast: rename variable into paced_train

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
editoast: rename 'items' into 'simulations'

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
editoast: rename 'occurrences' to 'occurrences_count'

'occurrences' is not a variable that contains some `TrainSchedule`,
but only a count of the number of `TrainSchedule` (occurrences).

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

woshilapin committed:
editoast: send correct 'start_time' for paced train conflicts

Asking for conflicts for a paced train is about sending all of the
occurrences of the paced train to core. The same 'start_time' was sent
for all the occurrences instead of getting the calculated 'start_time'
for each occurrence. 

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

younesschrifi reviewed (approved):
Thanks for the PR and for the fix !

leovalais commented on editoast/src/core/pathfinding.rs:
```diff
@@ -105,7 +105,7 @@ pub enum PathfindingCoreResult {
 
 /// A successful pathfinding result. This is also used for STDCM response.
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, ToSchema)]
+#[derive(Default, Serialize, Deserialize, Clone, Debug, PartialEq, ToSchema)]
```
nit: can we use `cfg_attr(test` to better indicate that it's just a shortcut we use in tests?

woshilapin replied:
Changed in 6ec2e277.

leovalais commented on editoast/src/models/paced_train.rs:
```diff
@@ -19,7 +19,7 @@ use super::Tags;
 use crate::models::train_schedule::TrainSchedule;
 
-#[derive(Debug, Clone, Model)]
+#[derive(Debug, Default, Clone, Model)]
```
nit: same about cfg_attr

woshilapin replied:
I chose not too because the corresponding `TrainSchedule` doesn't have it.

woshilapin replied:
OK, I tested to move all of them as `#cfg_attr(test, derive(Default))]` and it seems fine. I feel it's probably a better thing because I'm not sure it make sense to have a `Default` implementations for these (except, it's handy for tests). Changed in 6ec2e277.

leovalais reviewed (approved):
lgtm, thanks for fixing this and adding tests

bougue-pe commented on tests/tests/test_timetable.py:
```diff
@@ -166,6 +166,69 @@ def test_conflicts(
+        (
+            "PT1M",
+            set({"Spacing"}),
+        ),  # Every second, all occurrences will fight for space
```
```suggestion
        ),  # Every minute, all occurrences will fight for space
```

woshilapin replied:
Done in [1894a19](https://github.com/OpenRailAssociation/osrd/compare/6ec2e2776e0cb44b86e035d89ce32697642a57ed..1894a19de130e593ad6441fd361da25130252bb8).

bougue-pe commented on tests/tests/test_timetable.py:
```diff
@@ -166,6 +166,69 @@ def test_conflicts(
+        f"{EDITOAST_URL}timetable/{timetable_id}/paced_trains",
+        json=[paced_train_payload],
+    )
+    print(stopping_paced_train_response.text)
```
Not sure you want to keep this `print`.

woshilapin replied:
Done in [1894a19](https://github.com/OpenRailAssociation/osrd/compare/6ec2e2776e0cb44b86e035d89ce32697642a57ed..1894a19de130e593ad6441fd361da25130252bb8).

bougue-pe commented on tests/tests/test_timetable.py:
```diff
@@ -166,6 +166,69 @@ def test_conflicts(
+        "exceptions": [],
+    }
+
+    stopping_paced_train_response = requests.post(
```
Probably a left over from copy-paste:
```suggestion
    paced_train_response = requests.post(
```

woshilapin replied:
Done in [1894a19](https://github.com/OpenRailAssociation/osrd/compare/6ec2e2776e0cb44b86e035d89ce32697642a57ed..1894a19de130e593ad6441fd361da25130252bb8).

bougue-pe reviewed (approved):
Thanks!

woshilapin committed:
editoast: move some 'Default' implementation to 'cfg(test)'

Implementing `Default` on `TrainSchedule`, `Paced`, and
`PathfindingResultSuccess` don't make a lot of sense: what default
values means for these? However, it's a nice convenience for defining
tests, focusing on only the fields that makes sense in the test.

Signed-off-by: Jean SIMARD <woshilapin@tuziwo.info>

shenriotpro commented on tests/tests/test_timetable.py:
```diff
@@ -166,6 +166,68 @@ def test_conflicts(
+        ("PT15M", set()),  # Every half-hour, no conflict between each occurrences
+        (
+            "PT1M",
+            set({"Spacing"}),
```
nit: that's just `{"Spacing"}`

woshilapin replied:
Indeed 🤦 Fixed in [5b277b5](https://github.com/OpenRailAssociation/osrd/compare/fa1276fd12c0ae32d1898f333cf4d40d000f78e9..5b277b5f232dc10fbe4587d233b4853bf82189cc).

shenriotpro reviewed (approved):
thanks
